clients/drcachesim/tracer/tracer.cpp

```diff
@@ -1398,6 +1409,100 @@ event_delay_app_instruction(void *drcontext, void *tag, instrlist_t *bb, instr_t
     return DR_EMIT_DEFAULT;
 }
 
+
+typedef struct _func_metadata {
+    const trace_marker_type_t marker_retaddr;
+    const trace_marker_type_t marker_arg;
+    const trace_marker_type_t marker_retval;
+    const char *name;
+    const int num_args;
+} func_metadata;
+
+#define MALLOC_RETADDR TRACE_MARKER_TYPE_FUNC_MALLOC_RETADDR
+#define MALLOC_ARG     TRACE_MARKER_TYPE_FUNC_MALLOC_ARG
+#define MALLOC_RETVAL  TRACE_MARKER_TYPE_FUNC_MALLOC_RETVAL
+
+static const func_metadata instru_funcs[] = {
```

Comment:
We'll obviously need to expand the list (especially with all the operator new variants), but I think it's fine to leave that for later.

I'd remove do_malloc and malloc_impl because they sound like implementation details of particular allocators and it's not guaranteed that the arguments / return values are what we expect them to be.

Reply:
I ported the [do_malloc](https://github.com/DynamoRIO/drmemory/blob/master/common/alloc.c#L373) and [malloc_impl](https://github.com/DynamoRIO/drmemory/blob/master/common/alloc.c#L408) from drmemory, and they seem to have the same signature as malloc. Do you think it is safe enough to keep them here?

Reply:
No idea about malloc_impl (I wouldn't worry too much about windows outside of breaking DR there).

do_malloc() is from an old version of tcmalloc and doesn't exist any more even upstream (https://github.com/gperftools/gperftools/search?q=do_malloc&unscoped_q=do_malloc). Even if it did, I don't think we should've kept it. It was an internal helper routine called by malloc(), so if we trace both, we'd record the same call twice (unless we taught the tracer we're in the same context, but that sounds like unnecessary complexity).

Reply:
I lied, it still exists upstream [1]. Still, we should ignore it so we don't double record.

[1] https://github.com/gperftools/gperftools/blob/f47a52ce85c3d8d559aaae7b7a426c359fbca225/src/tcmalloc.cc#L1348

Reply:
OK, will remove those. I guess it is also good to keep it simple as much as possible in this PR.

Reply:
Done